extensions/amp-subscriptions/0.1/amp-subscriptions.js

```diff
@@ -531,16 +539,113 @@ export class SubscriptionService {
 
   /**
    * Unblock document based on grant state and selected platform
-   * @param {boolean=} doPlatformSelection
+   * @param {boolean=} shouldActivatePlatform
+   * @return {!Promise}
    * @private
    */
-  startAuthorizationFlow_(doPlatformSelection = true) {
-    this.platformStore_.getGrantStatus().then((grantState) => {
-      this.processGrantState_(grantState);
-      this.performPingback_();
+  startAuthorizationFlow_(shouldActivatePlatform = true) {
+    const grantStatusPromise = this.platformStore_.getGrantStatus();
+    const grantEntitlementPromise = this.platformStore_.getGrantEntitlement();
+    const promises = Promise.all([grantStatusPromise, grantEntitlementPromise]);
+
+    return promises.then((results) => {
+      const granted = results[0];
+      const entitlement = results[1];
+
+      // Create shortcut to continue authorization flow.
+      const continueAuthorizationFlow = () =>
+        this.handleGrantState_({granted, shouldActivatePlatform});
+
+      if (!this.metering_) {
+        // Move along. This article doesn't need AMP metering's logic.
+        continueAuthorizationFlow();
+        return;
+      }
+
+      // AMP metering's logic:
+      // - Granted?
+      //   - Yes.
+      //     - From AMP metering?
+      //       - Yes. Consume entitlements.
+      //       - No. Handle grant state normally.
+      //   - No.
+      //     - Have we fetched AMP metering entitlements before?
+      //       - Yes. Handle grant state normally.
+      //       - No.
+      //         - Do we have AMP metering state?
+      //           - Yes. Fetch metering entitlements.
+      //           - No. Show regwall.
+
+      const meteringPlatform = this.platformStore_.getPlatform(
+        this.metering_.platformKey
+      );
+
+      if (granted) {
+        const grantCameFromAmpMetering =
+          entitlement &&
+          entitlement.grantReason === GrantReason.METERING &&
+          entitlement.service === this.metering_.platformKey;
+
+        if (!grantCameFromAmpMetering) {
+          // Move along. AMP metering isn't responsible for this grant.
+          continueAuthorizationFlow();
+          return;
+        }
+
+        // Consume metering entitlements.
+        const finishAuthorizationFlow = () => {
+          this.handleGrantState_({
+            granted: true,
+            shouldActivatePlatform: false,
+          });
+        };
+        meteringPlatform.activate(
+          entitlement,
+          entitlement,
+          finishAuthorizationFlow
+        );
+        return;
+      }
+
+      if (this.metering_.entitlementsWereFetchedWithCurrentMeteringState) {
```

Comment:
What if a user became eligible for new reads, even though the `userState` is the same?

Reply:
@ChrisAntaki - are you able to confirm on this one?

Reply:
The `entitlementsWereFetchedWithCurrentMeteringState` isn't persisted beyond a single pageload. The next time a user loads an article, we'll fetch entitlements again for the current metering state

Reply:
Ok, thanks for confirming!

Reply:
Np!